### PR TITLE
Sorted external hosts using CVMFS Stratum-1s.

### PR DESCRIPTION
**Pull request by bbockelm**
The external data hosts are assumed to not be running the GeoAPI
service.  Instead, use the CVMFS Stratum-1 servers listed in CVMFS_URL
to sort the external data hostnames with the GeoAPI.

@DrDaveD - this is the change we discussed earlier today.  Can you take a look at it?

### CODE REVIEW

File: cvmfs/download.cc [hunk 1]
```diff
@@ -1739,6 +1739,30 @@ void DownloadManager::SetHostChain(const string &host_list) {
...
+  }
+
+  opt_host_chain_ = new vector<string>(host_list);
+  opt_host_chain_rtt_ =
+    new vector<int>(opt_host_chain_->size(), kProbeUnprobed);
+  // LogCvmfs(kLogDownload, kLogSyslog, "using host %s",
+  //          (*opt_host_chain_)[0].c_str());
+  pthread_mutex_unlock(lock_options_);
+}
+
```
Comment (DrDaveD) [inline cvmfs/download.cc]: The two SetHostChain() functions are almost identical, can't one be implemented by calling the other?

**Comment by DrDaveD** [comment]
Overall it looks reasonable to me.  I haven't tested it though, I hope you have.  In fact the 054-geoapi test should probably be expanded to test sorting of external servers.

**Comment by bbockelm** [comment]
Yup - it tests fine locally.  Looking at 054-geoapi, is there really any way to run it outside CERN?

**Comment by DrDaveD** [comment]
It was limited to CERN until I [fixed](https://github.com/cvmfs/cvmfs/commit/37c0d431da817f33251f6a8f6288fee052f3f302) it 3 weeks ago.  Now it runs for me at Fermilab and should run elsewhere.  I see that it skips one check outside of CERN, return code 20. 

I just ran 054-geoapi on a fermicloud machine (without this pull request included) and it got almost all the way through but failed on the last test, 39.  Ah, that is because it's not at CERN after all.  I'm not sure how I missed it before.  It's comparing the order from a DIRECT proxy to an externally available proxy at CERN; that stays the same order at CERN but opposite order at FNAL.  I'll submit a patch tomorrow to have 054-geoapi do all checks correctly at CERN or FNAL and skip 20 and 39 elsewhere, like is done in the 614-geoservice order checks.

**Comment by bbockelm** [comment]
If you're fixing up the tests, can you also map unl.edu properly?  Should be the same ordering as FNAL.

**Comment by jblomer** [comment]
@bbockelm @DrDaveD Would we actually see multiple external host URLs?  I though for data there is something like a global redirector which acts as a central entry point.

**Comment by DrDaveD** [comment]
> Would we actually see multiple external host URLs? I though for data there is something like a global redirector which acts as a central entry point.

I asked Brian that yesterday and he said that the redirector was upstream of the external URL servers; those servers still need to be geosorted on the client.

File: cvmfs/download.cc [hunk 2]
```diff
@@ -1955,6 +1979,87 @@ void DownloadManager::ProbeHosts() {
...
+  }
+
+  std::vector<std::string> host_chain;
+  GetHostInfo(&host_chain, NULL, NULL);
+
+  std::vector<std::string> server_dns_names;
+  server_dns_names.reserve(servers->size());
+  for (unsigned i = 0; i < servers->size(); ++i) {
+    std::string host = dns::ExtractHost((*servers)[i]);
+    server_dns_names.push_back(host.empty() ? (*servers)[i] : host);
```
Comment (jblomer) [inline cvmfs/download.cc]: Currently merging, I'm wondering if `host.empty() ?` is necessary.  Can we just push back the result from `ExtractHost`?

File: cvmfs/cvmfs.cc [hunk 3]
```diff
@@ -2643,7 +2643,10 @@ static int Init(const loader::LoaderExports *loader_exports) {
     download::DownloadManager::kSetProxyBoth);
   g_external_download_ready = true;
   if (use_geo_api) {
-    cvmfs::external_download_manager_->ProbeGeo();
+    std::vector<std::string> host_chain;
```
Comment (jblomer) [inline cvmfs/cvmfs.cc]: This runs the Geo-API twice even if there are no dedicated external servers.

**Comment by bbockelm** [comment]
Yup - OSG runs a network of about 5 large-scale proxy servers ("StashCache" - each has 10TB storage, 10Gbps) that are meant to serve as a front for the data federation of smaller VOs (in some cases, the "federation" may be in reality be a single NFS server).

So we do expect non-trivial list of EXTERNAL_URLs in this case.

File: cvmfs/cvmfs.cc (hunk 3 above)
Comment (bbockelm) [inline cvmfs/cvmfs.cc]: It shouldn't.  There's an early-exit in the case where the sorting is trivial (less than 2 servers).

File: cvmfs/download.cc (hunk 1 above)
Comment (bbockelm) [inline cvmfs/download.cc]: Yeah, that's an easy fix.

File: cvmfs/download.cc (hunk 2 above)
Comment (bbockelm) [inline cvmfs/download.cc]: It is necessary as `ProbeGeo` calls with a list of hostnames but the initialization of `external_download_manager_` calls with a list of URLs.

File: cvmfs/cvmfs.cc (hunk 3 above)
Comment (jblomer) [inline cvmfs/cvmfs.cc]: Ah, I think the problem is that if external is the same as the normal urls, there is no need to do the sorting twice.  I.e., if `CVMFS_EXTERNAL_URL`, the external download manager should just apply the already sorted list from the normal one.

File: cvmfs/download.cc [hunk 4]
```diff
@@ -1955,6 +1961,87 @@ void DownloadManager::ProbeHosts() {
...
+  }
+  std::string host_list = JoinStrings(server_dns_names, ",");
+
+  // Protect against concurrent access to prng_
```
Comment (jblomer) [inline cvmfs/download.cc]: There is a duplicate of this block in `ProbeGeo()`, which I think should be removed.

File: cvmfs/download.cc [hunk 5]
```diff
@@ -1955,6 +1961,87 @@ void DownloadManager::ProbeHosts() {
...
+        break;
+      }
+    } else {
+      LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
+               "GeoAPI request %s failed with error %d [%s]",
+               url.c_str(), result, Code2Ascii(result));
+    }
+  }
+  if (!success) {
+    LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
```
Comment (jblomer) [inline cvmfs/download.cc]: There is a duplicate of this log in `ProbeGeo()`, which I think should be removed.

File: cvmfs/cvmfs.cc (hunk 3 above)
Comment (bbockelm) [inline cvmfs/cvmfs.cc]: Ah, ok.  That's an easy fix then.

**Comment by bbockelm** [comment]
Ok, I think I got all the review comments fixed.

**Pull request merged**